### PR TITLE
Handle localness in special cases by checking exported variable assignment

### LINKED ISSUE #42976: [Typescript] Find All References does not find object properties in multi-module workspace
*TS Template added by @mjbvz*

**TypeScript Version**: 4.2.2.

**Search Terms**
- references
- tsconfig

---

- VS Code Version: 1.53.2
- OS Version: Darwin x64 18.7.0

In a multi-module/project workspace that's using Typescript [Project References](https://www.typescriptlang.org/docs/handbook/project-references.html), if a shared TS project exports an object like:

```ts
export const foo = {
  bar: () => {}
};
```

And two TS projects are consuming it, if you open the workspace in VSCode, the Find All References command invoked on `.bar()` will only show 2/3 references. Interestingly, if you invoke the command on `foo`, 3/3 references will be found. And if you were to have all 3 files open, the issue does not reproduce (which makes this issue suspiciously similar to https://github.com/microsoft/vscode/issues/80087).

In the workspace, all 3 projects have `"composite": true` set and there's a root `tsconfig.json` referencing all 3, per [official docs](https://www.typescriptlang.org/docs/handbook/project-references.html#overall-structure) (without the root `tsconfig.json`, Find All References won't work for `foo` either).

Steps to Reproduce:

1. Clone https://github.com/WestonThayer/bug-vscode-find-refs (minimal repro)
2. Open in VSCode
3. Open `/api/src/server.ts`
4. Right click `.bar()` and try the Find All References command

ACTUAL: only `api/src/server.ts` and `shared/src/index.ts` are found as references
EXPECTED: all references are found, just like they would be if you right clicked `foo`

Does this issue occur when all extensions are disabled?: Yes

## Notes

Only applies to when the key of a shared, exported object is a function. For example if the shared code had `export const foo = () => {};`, Find All References would work fine.

### PR DESCRIPTION
**Pull request by sheetalkamat**
Fixes #42976

### CODE REVIEW

**Review by andrewbranch — APPROVED** [review]
Note that this is the only place `checker.isDeclarationVisible` is used—seems like it could just be renamed to `isDefinitionVisible` and augmented with this code. Though, the distinction between the two is somewhat useful to understanding the code.

**Comment by sheetalkamat** [comment]
`isDeclarationVisible` is introduced to be used by declaration emitter and hence keeping it distinct.

**Comment by sheetalkamat** [comment]
Just noticed there are some test failures, i will take a look at them.

File: src/services/goToDefinition.ts [hunk 1]
```diff
@@ -326,10 +326,41 @@ namespace ts.GoToDefinition {
...
+            isLocal: !isDefinitionVisible(checker, declaration)
         };
     }
 
+    function isDefinitionVisible(checker: TypeChecker, declaration: Declaration): boolean {
+        if (checker.isDeclarationVisible(declaration)) return true;
+        if (!declaration.parent) return false;
+
+        // Assignments to variable are visible if variable is visible
+        if (hasInitializer(declaration.parent) && declaration.parent.initializer === declaration) return isDefinitionVisible(checker, declaration.parent as Declaration);
```
Comment (amcasey) [inline src/services/goToDefinition.ts:338]: This is for cases like `const c = function() { ... }`, where `isDefinitionVisible` is called on the function expression?

Comment (amcasey) [reply src/services/goToDefinition.ts:338]: Maybe "assignments to variables" would be clearer as "variable initializers"?

**Review by amcasey — COMMENTED** [review]
The bug mentions that file openness affects the behavior, but I don't see anything about that in this PR.  Am I missing something?

**Comment by sheetalkamat** [comment]
Depending on which project is open the set of things where it would be found is going to be different so its not relevant to this issue. The main issue is if project set would be opened based on localness (which was set to local before and now we try to make it not local) making it work.

**Review by amcasey — APPROVED** [review]
Thanks for the explanation!

File: src/services/goToDefinition.ts [hunk 2]
```diff
@@ -326,10 +326,41 @@ namespace ts.GoToDefinition {
...
-            isLocal: !checker.isDeclarationVisible(declaration)
+            isLocal: !isDefinitionVisible(checker, declaration)
         };
     }
 
+    function isDefinitionVisible(checker: TypeChecker, declaration: Declaration): boolean {
+        if (checker.isDeclarationVisible(declaration)) return true;
+        if (!declaration.parent) return false;
+
+        // Assignments to variable are visible if variable is visible
```
Comment (sheetalkamat) [inline src/services/goToDefinition.ts:337]:
```suggestion
        // Variable initializers are visible if variable is visible
```

**Pull request merged**